### DISCUSSION TITLE
Question

### DISCUSSION THREAD

**Discussion by Sakura-0407** [binsec/binsec#43] [Q&A]
```
.intel_syntax noprefix
    .globl Fr_toInt
Fr_toInt:
        mov     rax, [rdi]

        movsx   rax, eax
        ret

    .type Fr_toInt,@function
	.size Fr_toInt,. - Fr_toInt
```
```
rdi := nondet as src2
rsi := nondet as src1
rax := nondet as dst
rbx := nondet as is_false
eax := nondet as low_rax
         
assume src1 & 0b111 = 0
assume src2 & 0b111 = 0                

assume 0x3000 <= src1 <= 0x4000  ## arbitrary range  
assume 0x5000 <= src2 <= 0x6000  ## arbitrary range  


starting from <Fr_toInt> with
    #assume @[src1, 32]{255} = 0
    #assume @[src1, 32] = 0x274b96d329e3b4496847f1dc39601d16fc07ee5f0a8e418e2d8c8a81c6c46537
    assume @[src1, 40] = 0x00000000000000000000000000000000000000000000000000000000000000000000000000000001
    assume @[src2, 40] = 0x00000000000000000000000000000000000000000000000000000000000000000000000000000002

end


replace <Fr_toInt:last> by
   eq<1> := true
   #0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
    print @[src1, 40]
    print @[src2, 8]
    print low_rax
    print dst
  assert eq
  halt
end

explore all
```
Hello. Above is my compiled code and validation script, shouldn't it return low_rax equal 0x00000002 and dst equal 0x0000000000000002?

Why does binsec always return the following:
```
zeno@ubuntu:~/Desktop/binsec$ binsec -sse -sse-script Fr_toInt.ini Fr_toInt.o -fml-solver-timeout 0 -fml-solver bitwuzla
[sse:result] Value @[src1<64>,40] : 0x00000000000000000000000000000000000000000000000000000000000000000000000000000001
[sse:result] Value @[src2<64>,8] : 0x0000000000000002
[sse:result] Value low_rax<32> : 0x158ba9aa
[sse:result] Value dst<64> : 0x000000001e227301
[sse:info] Empty path worklist: halting ...
[sse:info] SMT queries
```
I would appreciate it if you could answer for me

update：
It seems that when I try to assign a value to rax, rbx, etc., in the compiled function, binsec does not output the value I give

**Comment by recoules** [comment]
Hi @MJJ-Shuai,

I think there is a misunderstanding about the role of some assignments.
Especially, the `as` do not create a "forward" alias, it just gives an intermediate name to the value.
So, for instance, the two following snippets are equivalent.
```text
eax := nondet as low_rax
```
```text
low_rax<32> := nondet
eax := low_rax
```

So, your example program is doing the following.
```text
dst<64> := nondet
rax := dst
low_rax<32> := nondet
rax := rax{63..32} :: low_rax
rax := @[rdi, 8]
rax := sext64 rax{31..0}
print low_rax
print dst
```

Since `low_rax` and `dst` do not have been constrained nor assigned again, `binsec` prints the initial (non-deterministic) values.
This is the intended behavior since it allows finding the initial input value that leads to a given program state. If you want to print the values at the end of the function, you have to use the actual location.
For instance, I think the following will print what you want.

```text
print eax
print rax
```

Regards,
Frédéric

**Reply by Sakura-0407** [reply]
Oh, I see. It seems that I misunderstood the coding rules of binsec. Thank you again.
